src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1022,250 +1015,222 @@ private ImmutableArray<BoundNode> ConvertCollectionExpressionElements(
+                            },
+                            (elementType, useIndexer, elementConversion, elementKeyValueTypes, collectionInitializerAddMethodBinder),
+                            diagnostics);
+                        break;
+                    case BoundKeyValuePairElement keyValuePairElement:
+                        convertedElement = convertKeyValuePair(
+                            this,
+                            keyValuePairElement,
+                            elementConversion,
+                            elementType,
+                            elementKeyValueTypes,
+                            useIndexer,
+                            implicitReceiver,
+                            collectionInitializerAddMethodBinder,
+                            diagnostics);
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(element);
                 }
+                builder.Add(convertedElement);
             }
-            else if (elementType is { } && ConversionsBase.TryGetCollectionKeyValuePairTypes(Compilation, collectionTypeKind, elementType) is (var elementKeyType, var elementValueType))
-            {
-                var elementConversions = conversion.UnderlyingConversions;
 
-                Debug.Assert(elementType is { });
-                Debug.Assert(elementConversions.All(c => c.Exists));
-                Debug.Assert(implicitReceiver is { });
+            Debug.Assert(conversionIndex == elementConversions.Length);
+            conversion.MarkUnderlyingConversionsChecked();
 
-                int conversionIndex = 0;
-                foreach (var element in elements)
+            return builder.ToImmutableAndFree();
+
+            static BoundNode convertKeyValuePair(
+                Binder binder,
+                BoundKeyValuePairElement keyValuePairElement,
+                Conversion elementConversion,
+                TypeSymbol elementType,
+                (TypeSymbol Key, TypeSymbol Value)? elementKeyValueTypes,
+                bool useIndexer,
+                BoundObjectOrCollectionValuePlaceholder? implicitReceiver,
+                CollectionInitializerAddMethodBinder? collectionInitializerAddMethodBinder,
+                BindingDiagnosticBag diagnostics)
+            {
+                if (elementKeyValueTypes is (var elementKeyType, var elementValueType) &&
+                    elementConversion.TryGetKeyValueConversions(out var keyConversion, out var valueConversion))
                 {
-                    if (element is BoundCollectionExpressionWithElement)
+                    var keyValuePairSyntax = (KeyValuePairElementSyntax)keyValuePairElement.Syntax;
+                    var keyValuePairConstructor = useIndexer ? null : (MethodSymbol?)binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__ctor, diagnostics, syntax: keyValuePairSyntax);
```

Comment:
Consider moving this inside the `if` below (ie. closer to usage). Otherwise, it looks like we unnecessarily check for the presence of this well-known member

Reply:
The `KeyValuePair<,>..ctor` is needed for lowering regardless of whether `Add()` is called, and this call to `GetWellKnownTypeMember()` is needed for reporting missing member, but the local variable is only needed for the `Add()` case.